### ISSUE TITLE
search: and/or queries performance drop

### ISSUE THREAD

**Issue by keegancsmith** [sourcegraph/sourcegraph#31400]
Around 2022-02-16 04:32:00 UTC the performance of and/or queries degraded. In particular when I look at the search blitz graphs for time to first result for `searchblitz_and_regex` and `searchblitz_or_regex`. I haven't investigated the cause further, but given the time this happened and the refactoring work going on in the frontend I assume this is related to one of those PRs going out.

Both times went up, but or queries seem much more affected.

| name               | before | after |
|------------------ |------ |----- |
| `search_blitz_and` | 0.11s  | 0.26s |
| `search_blitz_or`  | 0.09s  | 0.62s |

[Grafana link](https://sourcegraph.com/-/debug/grafana/explore?orgId=1&left=%5B%221644948141661%22,%221645090528079%22,%22Prometheus%22,%7B%22expr%22:%22sum(rate(src_search_streaming_latency_seconds_sum%7Bsource%3D~%5C%22searchblitz.*regex.*%5C%22,%20source!~%5C%22.*mono.*%5C%22%7D%5B1h30m0s%5D))%20by%20(source)%20%2F%20sum(rate(src_search_streaming_latency_seconds_count%7Bsource%3D~%5C%22searchblitz.*%5C%22%7D%5B1h30m0s%5D))%20by%20(source)%22,%22datasource%22:%22Prometheus%22,%22exemplar%22:true,%22requestId%22:%22Q-0b00a41e-6959-4db2-9351-0ade981fb9a9-0A%22%7D%5D)

![image](https://user-images.githubusercontent.com/187831/154488980-e9799907-ae74-45dc-8967-ab95ca61fd89.png)

cc @rvantonder @sourcegraph/search-core

**Comment by keegancsmith** [comment]
Did some further exploration in honeycomb and noticed `searchblitz_symbol_small` p90 also regressed around the same time. https://ui.honeycomb.io/sourcegraph/datasets/search/result/87UHQxpTk6d

**Comment by tsenart** [comment]
Cc @camdencheek

**Comment by camdencheek** [comment]
The `or` performance regression was expected. The new `Or` job doesn't stream until we get a match from each subquery so that we don't send incompletely matched documents. On top of this, the search blitz query we are using for `or` is a particularly nasty case that makes it so we don't send any matches until all suqueries finish.

The `and` and symbols change is unexpected though. I will look into that this morning. 

Also, the goal here is to very soon move these `and` and `or` terms into the zoekt query, which should dramatically improve performance

**Comment by camdencheek** [comment]
More discussion about the original decision [here](https://github.com/sourcegraph/sourcegraph/pull/30655#discussion_r799060519)

**Comment by camdencheek** [comment]
[PR](https://github.com/sourcegraph/sourcegraph/pull/31404) to add saner `or` query to search blitz to have a nicer point of comparison

**Comment by camdencheek** [comment]
Looking at the code, I expect the `and` regression is that, at the same time, we moved the query limit from below the `And` node (limit each subquery to 500 results) to above the `And` node (limit the intersection of subqueries to 500 results). It seems very possible this is accounting for the difference here. Do we track the number of results for our sentinel queries anywhere? This should be a visible increase in that number.

**Comment by chrismwendt** [comment]
This trail...

![CleanShot 2022-02-17 at 09 58 00@2x](https://user-images.githubusercontent.com/1387653/154532121-d98dca31-b814-406f-adbe-abfa7bed3d3f.png)

...led me to these 2 PRs:

- https://github.com/sourcegraph/sourcegraph/pull/31291
- https://github.com/sourcegraph/sourcegraph/pull/31286

By looking at this graph, it's not clear that the symbols search is slower:

<img width="1063" alt="CleanShot 2022-02-17 at 16 30 37@2x" src="https://user-images.githubusercontent.com/1387653/154589103-e2a007ad-6550-4cb6-815d-67c19eca436e.png">

I don't think any of my [recent commits](https://sourcegraph.com/search?q=context:global+repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+type:commit+author:Chris+after:2022-02-10+before:2022-02-16T04:36:00Z&patternType=literal) would have caused this slowness:

<img width="890" alt="CleanShot 2022-02-17 at 10 33 52@2x" src="https://user-images.githubusercontent.com/1387653/154538270-ca664cc7-5b72-4bd6-b463-63227aae6928.png">

**Comment by camdencheek** [comment]
Closing this since:
1) this was a conscious change that traded some latency for completeness,
2) the queries used by search-blitz are not representative of the average case, and are particularly sensitive to this change, and
3) in most cases, this is fixed by Rijnard's recent changes that send optimized `and` queries to Zoekt. You'll see the `and` latency [dropped](https://sourcegraph.com/-/debug/grafana/d/frontend/frontend?viewPanel=102731&from=now-2d&to=now) to below the original level

Happy to re-open if there are any remaining concerns

**Comment by keegancsmith** [comment]
> the queries used by search-blitz are not representative of the average case, and are particularly sensitive to this change, and

what can we do to search-blitz to make it more representative of average case so it is more useful?

**Comment by camdencheek** [comment]
Closing the loop here, I merged a PR today that applies the same optimization to `or` queries. It's a little early, but the initial metrics seem to have stabilized around 90ms mean query latency for `searchblitz_or_regex`, which is back where it started. 

Additionally, I opened [a PR](https://github.com/sourcegraph/sourcegraph/pull/34486) that adds an `and_literal_simple` searchblitz query to mirror the `or_literal_simple`. This will allow us to measure the overhead of an `and` query in its simplest form. Right now, the only `and` query we've got is a 4-term regex one, which is valuable, but not a great representation of normal `and` queries.
